### ISSUE TITLE
`IndexError`s are handled differently than other exceptions

### ISSUE THREAD

**Issue by AugustKarlstedt** [pyparsing/pyparsing#573]
When a parse action raises an `IndexError`, it does not bubble up the exception message as it would other types of exceptions. This makes it hard to debug how/why my parse action is failing if it raises an `IndexError`.

minimal example:

```python
import pyparsing as pp

def parse_hello(
    _string: str, _location: int, _tokens: pp.ParseResults
) -> None:
    print(f"parsed {_tokens}")
    raise Exception('this works ok')

def parse_world(
    _string: str, _location: int, _tokens: pp.ParseResults
) -> None:
    print(f"parsed {_tokens}")
    raise IndexError('this does not')

hello = pp.Keyword("hello").set_parse_action(parse_hello)
world = pp.Keyword("world").set_parse_action(parse_world)
parser = pp.ZeroOrMore(hello | world)

parser.run_tests([
    "hello",
    "world"
])
```

outputs

```
parsed ['hello']

hello
FAIL-EXCEPTION: Exception: this works ok
parsed ['world']

world
world
^
ParseException: Expected end of text, found 'world'  (at char 0), (line:1, col:1)
FAIL: Expected end of text, found 'world'  (at char 0), (line:1, col:1)
```

which properly bubbles up the exception raised when parsing "hello" but fails with an ambiguous error when parsing "world". 

It should print `FAIL-EXCEPTION: IndexError: this does not` even if the parse action raises an `IndexError`

**Comment by ptmcg** [comment]
Thanks - I actually had some code around this as being expected behavior, but giving it a re-think, I worked out how to keep parse action IndexErrors from getting misread as ParserElement.parseImpl IndexErrors. In the bargain, I also enhanced the run_tests output to now indicate if an exception is raised, if it is raised in a parse action, and reports the parse action by name.  For your case, I now get this output:

```
parsed ['hello']


hello
FAIL-EXCEPTION (raised in parse action 'parse_hello'): Exception: this works ok
parsed ['world']


world
FAIL-EXCEPTION (raised in parse action 'parse_world'): IndexError: this does not
```

This fix will go out in the next release.

**Comment by ptmcg** [comment]
Please try again using 3.2.0rc1, which I just pushed to PyPI.

**Comment by AugustKarlstedt** [comment]
Just tested in a non-trivial codebase and the error message now shows exactly where the underlying `IndexError` happens. Much appreciated for the improvements here!

**Comment by ptmcg** [comment]
Released in 3.2.0, closing.
